waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -294,37 +318,41 @@
           pod-started-at (-> pod (get-in [:status :startTime]) timestamp-str->datetime)
-          {:keys [waiter/revision-timestamp]} (get-in pod [:metadata :annotations])]
-      (scheduler/make-ServiceInstance
-        (cond-> {:extra-ports (->> pod-annotations :waiter/port-count Integer/parseInt range next (mapv #(+ port0 %)))
-                 :flags (cond-> #{}
-                          (check-expired scheduler service-id instance-id pod-restart-count pod-annotations primary-container-status pod-started-at)
-                          (conj :expired))
-                 :healthy? (true? (get primary-container-status :ready))
-                 :host (get-in pod [:status :podIP] scheduler/UNKNOWN-IP)
-                 :id instance-id
-                 :k8s/api-server-url api-server-url
-                 :k8s/app-name (get-in pod [:metadata :labels :app])
-                 :k8s/namespace (k8s-object->namespace pod)
-                 :k8s/pod-name (k8s-object->id pod)
-                 :k8s/restart-count primary-container-restart-count
-                 :k8s/user run-as-user
-                 :log-directory (log-dir-path run-as-user primary-container-restart-count)
-                 :port port0
-                 :service-id service-id
-                 :started-at pod-started-at}
-          node-name (assoc :k8s/node-name node-name)
-          phase (assoc :k8s/pod-phase phase)
-          revision-timestamp (assoc :k8s/revision-timestamp revision-timestamp)
-          (seq container-statuses) (assoc :k8s/container-statuses
-                                          (map (fn [{:keys [state] :as status}]
-                                                 (when (> (count state) 1)
-                                                   (log/warn "received multiple states for container:" status))
-                                                 (let [[state {:keys [reason]}] (first state)]
-                                                   (cond-> (select-keys status [:image :name :ready])
-                                                     (some? reason) (assoc :reason reason)
-                                                     (some? state) (assoc :state state))))
-                                               container-statuses)))))
+          {:keys [waiter/revision-timestamp]} (get-in pod [:metadata :annotations])
+          instance (scheduler/make-ServiceInstance
+                     (cond-> {:extra-ports (->> pod-annotations :waiter/port-count Integer/parseInt range next (mapv #(+ port0 %)))
+                              :flags (cond-> #{}
+                                       (check-expired scheduler service-id instance-id pod-restart-count pod-annotations primary-container-status pod-started-at)
+                                       (conj :expired))
+                              :healthy? (and (true? (get primary-container-status :ready))
+                                             (not frequently-restarting?))
```

Comment:
Why are we changing the logic for `healthy?`? I think we eject instances before killing them when we do a normal scale-down operation, so shouldn't we do the same thing here?

Reply:
Since we mark the instance _unhealthy_ as soon as we see the container restart count go up, it doesn't have a chance to actually become _ready_, so I don't think there's no chance of a race across the routers. Based on that plus the fact that it sounds like getting Waiter to do ejection in this case (via the auto-scaler) sounds very complex, I guess I'm OK with this for now. Maybe just add a comment like the following above the new clause on line 328:
```
;; Note that when frequently-restarting? becomes true, the container *just* restarted and is non-ready,
;; therefore it's impossible to observe a healthy? status between the time the container restarted
;; and the time that the instance was marked permanently unhealthy due to the high restart count.
```